client/src/main/java/com/vesoft/nebula/client/meta/MetaManager.java

```diff
@@ -0,0 +1,498 @@
+                    metaClient.connect();
+                }
+            }
+        }
+    }
+
+    /**
+     * init the meta info cache
+     * make sure this method is called before use metaManager
+     */
+    public void init() {
+        fillMetaInfo();
+    }
+
+    /**
+     * close meta client
+     */
+    public void close() {
+        metaClient.close();
+    }
+
+
+    /**
+     * fill the meta info
+     */
+    private synchronized void fillMetaInfo() {
+        for (IdName space : metaClient.listSpaces()) {
+            // space schema
+            String spaceName = new String(space.getName());
+            metaInfo.getSpaceNameMap().put(spaceName, space.getId().getSpace_id());
+            metaInfo.getSpacePartLocation().put(spaceName, metaClient.getPartsLocation(spaceName));
+
+            // tag schema
+            freshTags(spaceName);
+
+            // edge schema
+            freshEdges(spaceName);
+        }
+    }
+
+
+    /**
+     * get space id
+     *
+     * @param spaceName nebula space name
+     * @return
+     */
+    public int getSpaceId(String spaceName) {
+        if (!metaInfo.getSpaceNameMap().containsKey(spaceName)) {
+            fillMetaInfo();
+        }
+        if (!metaInfo.getSpaceNameMap().containsKey(spaceName)) {
+            throw new IllegalArgumentException("space does not exist");
+        }
+        return metaInfo.getSpaceNameMap().get(spaceName);
+    }
+
+
+    /**
+     * get tag id
+     *
+     * @param spaceName nebula graph space name
+     * @param tagName   nebula tag name
+     * @return long
+     */
+    private long getTagId(String spaceName, String tagName) {
+        if (!metaInfo.getTagNameMap().containsKey(spaceName)
+                || !metaInfo.getTagNameMap().get(spaceName).containsKey(tagName)) {
+            fillMetaInfo();
+        }
+        Map<String, Map<String, Long>> tagNameMap = metaInfo.getTagNameMap();
+        if (!tagNameMap.containsKey(spaceName)) {
+            throw new IllegalArgumentException("space does not exist.");
+        }
+
+        if (!tagNameMap.get(spaceName).containsKey(tagName)) {
+            throw new IllegalArgumentException(String.format("tag %s does not exist in space %s",
+                    tagName, spaceName));
+        }
+        return tagNameMap.get(spaceName).get(tagName);
+    }
+
+    /**
+     * get Tag
+     *
+     * @param spaceName nebula space name
+     * @param tagName   nebula tag name
+     * @return
+     */
+    public TagItem getTag(String spaceName, String tagName) {
+        long tagId = getTagId(spaceName, tagName);
+        return metaInfo.getSpaceTagItems().get(spaceName).get(tagId);
+    }
+
+
+    /**
+     * get edge id
+     *
+     * @param spaceName nebula graph space name
+     * @param edgeName  nebula edge name
+     * @return long
+     */
+    private long getEdgeId(String spaceName, String edgeName) {
+        if (!metaInfo.getTagNameMap().containsKey(spaceName)
```

Comment:
the `read` also needs to add a lock. you can use the lock.readLock()

Reply:
read does not need to lock, only write cause map's endless loop

Reply:
the map's value is the list, the list will change, such as the schema's columns. it means you read the list and rewrite the list.

Reply:
if It happens，it means the map‘s value need to be update.

Reply:
ok,  Java's garbage collection mechanism makes it safe. But it's still not safe. if the key not exist, here will multi-thread call fillMetaInfo(), and the fillMetaInfo() is not thread-safe.

Reply:
fillMetaInfo() is thread safe, this method is synchronized.